src/coreclr/src/pal/src/memory/pressure.cpp

```diff
@@ -0,0 +1,177 @@
+        return -1;
+    }
+
+    fdHighMemTrigger = CreatePSIMemTrigger("some", 5000, 1000000);
+    if (fdHighMemTrigger < 0)
+    {
+        goto closeLowMemTriggerFd;
+    }
+
+    while (TRUE)
+    {
+        struct pollfd fds[] =
+        {
+            {fdLowMemTrigger, POLLPRI, 0},
+            {fdHighMemTrigger, POLLPRI, 0},
+        };
+        int nFds = poll(fds, 2, -1);
+
+        if (nFds < 0)
+        {
+            if (errno == EINTR)
+            {
+                continue;
+            }
+
+            break;
+        }
+
+        for (int i = 0; i < nFds; i++)
+        {
+            if (fds[i].revents & POLLERR)
+            {
+                continue;
+            }
+
+            if (fds[i].revents & POLLPRI)
+            {
+                if (fds[i].fd == fdLowMemTrigger)
+                {
+                    SetEvent(notificationHandles[LowMemoryResourceNotification]);
+                }
+                else if (fds[i].fd == fdHighMemTrigger)
+                {
+                    SetEvent(notificationHandles[HighMemoryResourceNotification]);
+                }
+            }
+        }
+    }
+
+    close(fdHighMemTrigger);
+
+closeLowMemTriggerFd:
+    close(fdLowMemTrigger);
+    return -1;
+}
+
+static BOOL CreateEventForType(MEMORY_RESOURCE_NOTIFICATION_TYPE type)
+{
+    static const LPCWSTR eventNamesByType[] =
+    {
+        W("LowMemoryResourceNotification"),
+        W("HighMemoryResourceNotification"),
+    };
+
+    // Instead of being a HANDLE created by CreateEvent(), this would
+    // ideally be something managed by the PAL in order to properly wrap the
+    // trigger file descriptor, and the polling thread.  This way, if
+    // CloseHandle() is called on this handle, its associated trigger file
+    // descriptor would also be closed.
+    //
+    // SetHandleInformation() could be optionally be called in this HANDLE
+    // to set HANDLE_FLAG_PROTECT_FROM_CLOSE, but it's not implemented by
+    // the PAL at the moment.  The only place where
+    // CreateMemoryResourceNotification() is used in the runtime is in the
+    // VM, and it doesn't call CloseHandle() for this notification type.
+
+    notificationHandles[type] = CreateEventW(/* lpEventAttributes */ nullptr,
+                                             /* bManualReset */ FALSE,
+                                             /* bInitialState */ FALSE,
+                                             eventNamesByType[type]);
+
+    return notificationHandles[type] != nullptr;
+}
+
+HANDLE CreateMemoryResourceNotification(MEMORY_RESOURCE_NOTIFICATION_TYPE notificationType)
+{
+    if (pollingThread == nullptr)
+    {
+        if (!CreateEventForType(LowMemoryResourceNotification))
+        {
+            return nullptr;
+        }
+        if (!CreateEventForType(HighMemoryResourceNotification))
+        {
+            goto closeLowMemoryNotificationHandle;
+        }
+
+        pollingThread = CreateThread(/* lpThreadAttributes */ nullptr,
```

Comment:
Can we use `CPalSynchronizationManager::WorkerThread` for this? We have way too many threads in the system that do very little. It would be nice to not introduce yet another one.

Reply:
Will take a look at that; was not aware of it.

Reply:
I don't really see how I can use that to monitor these file descriptors, and I can't find any facility to monitor them, either (and the canonical way to poll for events in PSI requires looking for the POLLPRI flag, etc.).

Reply:
I wonder if we could integrate it into the code in CPalSynchronizationManager::ReadBytesFromProcessPipe. That's the thing running on the thread @jkotas has mentioned and that's waiting for stuff from the pipe using poll. Maybe we could add the pressure polling there (after a little refactoring).